File: server/src/game/net/Headers.ts

export default class Header {

	readonly code: number;
	private constructor(code: number) {
		this.code = code;
	}

	static from(code: number) {
		return new Header(code);
	}

	//#region CRUD Operations
	public static readonly CREATE = 0x1;
	public static readonly LOOKUP = 0x2;
	public static readonly UPDATE = 0x3;
	public static readonly DELETE = 0x4;
	//#endregion

	//#region Network
	public static readonly NETWORK = 0x1000;
	
	public static readonly SERVER = Header.NETWORK | 0x100;
	public static readonly ServerSaving = Header.from(Header.SERVER | 0x01);
	public static readonly ServerShutdown = Header.from(Header.SERVER | 0x02);
	public static readonly ServerMessage = Header.from(Header.SERVER | 0x03);
	
	public static readonly CLIENT = Header.NETWORK | 0x200;
	public static readonly BindClientID = Header.from(Header.CLIENT | 0x01);
	public static readonly Authenticate = Header.from(Header.CLIENT | 0x02);
	//#endregion

	//#region Object Messages
	public static readonly OBJECT = 0x2000;
	public static readonly PLAYER = Header.OBJECT | 0x200;
	public static readonly MOBILE = Header.OBJECT | 0x400;

	public static readonly ObjectCreate = Header.from(Header.OBJECT | Header.CREATE);
	public static readonly ObjectUpdate = Header.from(Header.OBJECT | Header.UPDATE);
	public static readonly ObjectDelete = Header.from(Header.OBJECT | Header.DELETE);

	public static readonly PlayerUpdate = Header.from(Header.PLAYER | Header.UPDATE);

	public static readonly MobileCreate = Header.from(Header.MOBILE | Header.CREATE);
	public static readonly MobileUpdate = Header.from(Header.MOBILE | Header.UPDATE);
	public static readonly MobileDelete = Header.from(Header.MOBILE | Header.DELETE);
	//#endregion

	public static readonly UNUSED_4 = 0x4000;
	public static readonly UNUSED_8 = 0x8000;

	//#region Errors
	public static readonly ERROR = 0xF000;
	public static readonly Unauthorized = Header.from(Header.ERROR | 0x401);
	public static readonly Forbidden = Header.from(Header.ERROR | 0x403);
	public static readonly NotFound = Header.from(Header.ERROR | 0x404);
	//#endregion
}

File: client/Assets/Scripts/Net/Socket/Headers.cs
namespace Assets.Scripts.Net
{
	public struct Header
	{
		private readonly ushort code;
		private Header(ushort code)
		{
			this.code = code;
		}

		public bool HasFlags(ushort flags) => (code & flags) != 0;

		public static implicit operator Header(int code) => new((ushort)code);
		public static implicit operator Header(uint code) => new((ushort)code);
		public static implicit operator Header(ushort code) => new(code);
		public static implicit operator int(Header header) => header.code;
		public static implicit operator uint(Header header) => header.code;
		public static implicit operator ushort(Header header) => header.code;

		public const ushort INVALID = 0x0;

		#region CRUD Operations
		public const ushort CREATE = 0x1;
		public const ushort LOOKUP = 0x2;
		public const ushort UPDATE = 0x3;
		public const ushort DELETE = 0x4;
		#endregion

		#region Network
		public const ushort NETWORK = 0x1000;

		public const ushort SERVER = NETWORK | 0x100;
		public const ushort ServerSaving = SERVER | 0x01;
		public const ushort ServerShutdown = SERVER | 0x02;
		public const ushort ServerMessage = SERVER | 0x03;

		public const ushort CLIENT = NETWORK | 0x200;
		public const ushort BindClientID = CLIENT | 0x01;
		public const ushort Authenticate = CLIENT | 0x02;
		#endregion

		#region Object Messages
		public const ushort OBJECT = 0x2000;
		public const ushort PLAYER = OBJECT | 0x200;
		public const ushort MOBILE = OBJECT | 0x400;

		public const ushort ObjectCreate = OBJECT | CREATE;
		public const ushort ObjectUpdate = OBJECT | UPDATE;
		public const ushort ObjectDelete = OBJECT | DELETE;

		public const ushort PlayerUpdate = PLAYER | UPDATE;

		public const ushort MobileCreate = MOBILE | CREATE;
		public const ushort MobileUpdate = MOBILE | UPDATE;
		public const ushort MobileDelete = MOBILE | DELETE;
		#endregion

		public const ushort UNUSED_4 = 0x4000;
		public const ushort UNUSED_8 = 0x8000;

		#region Errors
		public const ushort ERROR = 0xF000;
		public const ushort Unauthorized = ERROR | 0x401;
		public const ushort Forbidden = ERROR | 0x403;
		public const ushort NotFound = ERROR | 0x404;
		#endregion
	}
}
